server/src/main/java/org/opensearch/common/compress/ZstdCompressor.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.compress;
+
+import com.github.luben.zstd.RecyclingBufferPool;
+import com.github.luben.zstd.ZstdInputStreamNoFinalizer;
+import com.github.luben.zstd.ZstdOutputStreamNoFinalizer;
+import org.opensearch.common.bytes.BytesReference;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Arrays;
+
+/**
+ * {@link Compressor} implementation based on the ZSTD compression algorithm.
+ *
+ * @opensearch.internal
+ */
+public class ZstdCompressor implements Compressor {
+    // An arbitrary header that we use to identify compressed streams
+    // It needs to be different from other compressors and to not be specific
+    // enough so that no stream starting with these bytes could be detected as
+    // a XContent
+    private static final byte[] HEADER = new byte[] { 'Z', 'S', 'T', 'D', '\0' };
+    // 3 is a good trade-off between speed and compression ratio
+    private static final int LEVEL = 3;
+    // We use buffering on the input and output of in/def-laters in order to
+    // limit the number of JNI calls
+    private static final int BUFFER_SIZE = 4096;
+
+    private static final int ZSTD_BLOCK_LENGTH = 10 * 48 * 1024;
+
+    private static final int ZSTD_COMPRESSION_LEVEL = 3;
+
+    private static final int ZSTD_DICT_SIZE_FACTOR = 6;
+
+    @Override
+    public boolean isCompressed(BytesReference bytes) {
+        if (bytes.length() < HEADER.length) {
+            return false;
+        }
+        for (int i = 0; i < HEADER.length; ++i) {
+            if (bytes.get(i) != HEADER[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public int headerLength() {
+        return HEADER.length;
+    }
+
+    @Override
+    public InputStream threadLocalInputStream(InputStream in) throws IOException {
+        final byte[] header = in.readNBytes(HEADER.length);
+        if (Arrays.equals(header, HEADER) == false) {
+            throw new IllegalArgumentException("Input stream is not compressed with ZSTD!");
+        }
+        return new BufferedInputStream(new ZstdInputStreamNoFinalizer(in, RecyclingBufferPool.INSTANCE), BUFFER_SIZE);
```

Comment:
So I am unsure we need to wrap `ZstdInputStreamNoFinalizer` into `BufferedInputStream` ? (because it is already using buffers internally), may be we need 

```
new ZstdInputStreamNoFinalizer(new BufferedInputStream(in, BUFFER_SIZE), RecyclingBufferPool.INSTANCE));
```

? (also suggested here https://github.com/luben/zstd-jni/issues/194#issuecomment-974647924)

Reply:
Yeap. I will change it